cli/src/complete.rs

```diff
@@ -0,0 +1,122 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use clap::FromArgMatches as _;
+use clap_complete::CompletionCandidate;
+use jj_lib::workspace::DefaultWorkspaceLoaderFactory;
+use jj_lib::workspace::WorkspaceLoaderFactory as _;
+
+use crate::cli_util::expand_args;
+use crate::cli_util::find_workspace_dir;
+use crate::cli_util::GlobalArgs;
+use crate::command_error::internal_error;
+use crate::command_error::user_error;
+use crate::command_error::CommandError;
+use crate::config::default_config;
+use crate::config::LayeredConfigs;
+use crate::ui::Ui;
+
+pub fn local_bookmarks() -> Vec<CompletionCandidate> {
+    with_jj(|mut jj| {
+        jj.arg("bookmark")
+            .arg("list")
+            .arg("--template")
+            .arg(r#"if(!remote, name ++ "\n")"#)
+            .output()
+            .map(|output| output.stdout)
+            .map_err(user_error)
+            .and_then(|stdout| String::from_utf8(stdout).map_err(internal_error))
+            .map(|stdout| stdout.lines().map(CompletionCandidate::new).collect())
+    })
+}
+
+/// Shell out to jj during dynamic completion generation
+///
+/// In case of errors, print them and early return an empty vector.
+fn with_jj<F>(completion_fn: F) -> Vec<CompletionCandidate>
+where
+    F: FnOnce(std::process::Command) -> Result<Vec<CompletionCandidate>, CommandError>,
+{
+    get_jj_command()
+        .and_then(completion_fn)
+        .unwrap_or_else(|e| {
+            eprintln!("{}", e.error);
+            Vec::new()
+        })
+}
+
+/// Shell out to jj during dynamic completion generation
+///
+/// This is necessary because dynamic completion code needs to be aware of
+/// global configuration like custom storage backends. Dynamic completion
+/// code via clap_complete doesn't accept arguments, so they cannot be passed
+/// that way. Another solution would've been to use global mutable state, to
+/// give completion code access to custom backends. Shelling out was chosen as
+/// the preferred method, because it's more maintainable and the performance
+/// requirements of completions aren't very high.
+fn get_jj_command() -> Result<std::process::Command, CommandError> {
+    let current_exe = std::env::current_exe().map_err(user_error)?;
+    let mut command = std::process::Command::new(current_exe);
+
+    // Snapshotting could make completions much slower in some situations
+    // and be undesired by the user.
+    command.arg("--ignore-working-copy");
+
```

Comment:
nit: it's probably safer to add `--color=never` and `--no-pager`. `ui.color=always` can be set by user config.